File: backend/testcases/python/hard.py
def wordBreak(s, wordDict):
    word_set = set(wordDict)
    memo = {}

    def dfs(s):
        if s in memo:
            return memo[s]
        if not s:
            return [""]

        res = []
        for word in word_set:
            if s.startswith(word):
                sublist = dfs(s[len(word):])
                for sub in sublist:
                    if sub:
                        res.append(word + " " + sub)
                    else:
                        res.append(word)
        memo[s] = res
        return res

    return dfs(s)

# Test cases for wordBreak
try:
    result1 = wordBreak("catsanddog", ["cat","cats","and","sand","dog"])
    result2 = wordBreak("pineapplepenapple", ["apple","pen","applepen","pine","pineapple"])
    result3 = wordBreak("catsandog", ["cats","dog","sand","and","cat"])

    assert sorted(result1) == sorted(["cats and dog", "cat sand dog"])
    assert sorted(result2) == sorted(["pine apple pen apple", "pineapple pen apple", "pine applepen apple"])
    assert sorted(result3) == sorted([])

    print("ALL_TESTS_PASSED")
except:
    print("FAIL")

File: backend/testcases/java/hard.java


public class Main {
    public static boolean listsEqual(List<String> list1, List<String> list2) {
        return new HashSet<>(list1).equals(new HashSet<>(list2));
    }

    public static void main(String[] args) {
        try {
            WordBreak wb = new WordBreak(); // ✅ Create an instance

            List<String> result1 = wb.wordBreak("catsanddog", Arrays.asList("cat", "cats", "and", "sand", "dog"));
            List<String> expected1 = Arrays.asList("cats and dog", "cat sand dog");

            List<String> result2 = wb.wordBreak("pineapplepenapple", Arrays.asList("apple", "pen", "applepen", "pine", "pineapple"));
            List<String> expected2 = Arrays.asList("pine apple pen apple", "pineapple pen apple", "pine applepen apple");

            List<String> result3 = wb.wordBreak("catsandog", Arrays.asList("cats", "dog", "sand", "and", "cat"));
            List<String> expected3 = Collections.emptyList();

            assert listsEqual(result1, expected1) : "Test 1 Failed";
            assert listsEqual(result2, expected2) : "Test 2 Failed";
            assert listsEqual(result3, expected3) : "Test 3 Failed";

            System.out.println("ALL_TESTS_PASSED");
        } catch (AssertionError e) {
            System.out.println("FAIL");
        }
    }
}
